File: src/domain/score/factories/ProfilerChainFactory.js
const RiskQuizProfiler = require('../profilers/RiskQuizProfiler');
const IncomeProfiler = require('../profilers/IncomeProfiler');
const HouseProfiler = require('../profilers/HouseProfiler');
const VehicleProfiler = require('../profilers/VehicleProfiler');
const DependentProfiler = require('../profilers/DependentProfiler');
const MaritalStatusProfiler = require('../profilers/MaritalStatusProfiler');
const AgeProfiler = require('../profilers/AgeProfiler');

class ProfilerChainFactory {
  static getChain(profilerChain = 'default') {
    const availableChains = {
      default: [
        RiskQuizProfiler,
        VehicleProfiler,
        IncomeProfiler,
        HouseProfiler,
        DependentProfiler,
        MaritalStatusProfiler,
        AgeProfiler,
      ],
    };

    const chainNodes = availableChains[profilerChain];

    const TailNode = chainNodes.pop();

    let chain = new TailNode();

    chainNodes.reverse().forEach(Node => {
      chain = new Node(chain);
    });

    return chain;
  }
}

module.exports = ProfilerChainFactory;

File: src/domain/insurance_advice/factories/profiler_chain_factory.py
from src.domain.insurance_advice.profilers.house_profiler import HouseProfiler
from src.domain.insurance_advice.profilers.income_profiler import IncomeProfiler


def _get_nodes_by_chain(chain):
    available_chains = {
        'default': [HouseProfiler, IncomeProfiler]
    }

    return available_chains[chain]


def get_profiler_chain(profiler_chain='default'):
    chain_nodes = _get_nodes_by_chain(profiler_chain)

    TailNode = chain_nodes.pop()
    chain = TailNode()

    for Profiler in chain_nodes:
        chain = Profiler(chain)

    return chain
